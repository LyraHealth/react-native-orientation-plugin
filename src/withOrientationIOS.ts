import { withDangerousMod, ExportedConfigWithProps, WarningAggregator, withAppDelegate } from '@expo/config-plugins'
import { addLines, readFileAsync, saveFileAsync } from './utils'
import { ConfigPlugin } from '@expo/config-plugins'

import path from 'path'

export const addToAppDelegate = (src: string) => {
  const supportSrc = [
    `- (UIInterfaceOrientationMask)application:(UIApplication *)application supportedInterfaceOrientationsForWindow:(UIWindow *)window {
    return [Orientation getOrientation];
  }`
  ]
  src = addLines(src, '@implementation AppDelegate', 1, supportSrc)
  const lines = src.split('\n')
  lines.splice(1, 0, '#import "Orientation.h"')
  src = lines.join('\n')
  return src
}

const withOrientationAppDelegate: ConfigPlugin = (config) => {
  return withAppDelegate(config, config => {
    config.modResults.contents = addToAppDelegate(config.modResults.contents)
    return config
  })
}

const editPodfile = async (config: ExportedConfigWithProps, action: (podfile: string) => string) => {
  const podfilePath = path.join(config.modRequest.platformProjectRoot, 'Podfile')
  try {
    const podfile = action(await readFileAsync(podfilePath))
    return await saveFileAsync(podfilePath, podfile)
  } catch (e) {
    WarningAggregator.addWarningIOS('react-native-orientation-plugin', `Couldn't modify AppDelegate.m - ${e}.`)
  }
}

export const addToPodFile = (src: string, relativePath: string) => {
  // Match both variations of Ruby config:
  // unknown: pod 'react-native-orientation', path: '../node_modules/react-native-orientation'
  if (
    !src.match(/pod ['"]react-native-orientation['"],\s?path: ['"][^'"]*node_modules\/react-native-orientation['"]/)
  ) {
    src = addLines(src, 'use_react_native', 0, [`  pod 'react-native-orientation', path: '${relativePath}'`])
  }
  return src
}

const withOrientationPodFile: ConfigPlugin = (config) => {
  return withDangerousMod(config, [
    'ios',
    async config => {
      await editPodfile(config, podfile => {
        const packagePath = path.dirname(
          require.resolve(config.modRequest.projectRoot + '/node_modules/react-native-orientation/package.json')
        )
        const relativePath = path.relative(config.modRequest.platformProjectRoot, packagePath)
        podfile = addToPodFile(podfile, relativePath)
        return podfile
      })
      return config
    }
  ])
}

export const withOrientationIOS: ConfigPlugin = (config) => {
  config = withOrientationPodFile(config)
  config = withOrientationAppDelegate(config)
  return config
}
