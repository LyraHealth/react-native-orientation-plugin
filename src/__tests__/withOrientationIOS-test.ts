import { addToPodFile, addToAppDelegate } from '../withOrientationIOS'
import fs from 'fs'
import path from 'path'

describe(addToPodFile, () => {
  it(`modifies the PodFile file for react-native-orientation-plugin`, () => {
    const fixture = fs.readFileSync(path.join(__dirname, 'fixtures', 'PodFile'), 'utf8')
    expect(addToPodFile(fixture, '../node_modules/react-native-orientation')).toMatchSnapshot()
  })
})

describe(addToAppDelegate, () => {
  it('modifies the AppDelegate for react-native-orientation-plugin', () => {
    const fixture = fs.readFileSync(path.join(__dirname, 'fixtures', 'AppDelegate.m'), 'utf8')
    expect(addToAppDelegate(fixture)).toMatchSnapshot()
  })
})
