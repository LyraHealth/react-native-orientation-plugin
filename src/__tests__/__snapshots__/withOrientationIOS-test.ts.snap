// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addToAppDelegate modifies the AppDelegate for react-native-orientation-plugin 1`] = `
"#import \\"AppDelegate.h\\"
#import \\"Orientation.h\\"
#import <RNBranch/RNBranch.h>

#if defined(EX_DEV_MENU_ENABLED)
@import EXDevMenu;
#endif

#if defined(EX_DEV_LAUNCHER_ENABLED)
#include <EXDevLauncher/EXDevLauncherController.h>
#import <EXUpdates/EXUpdatesDevLauncherController.h>
#endif
#import <EXScreenOrientation/EXScreenOrientationViewController.h>

#import <React/RCTBridge.h>
#import <React/RCTBundleURLProvider.h>
#import <React/RCTRootView.h>
#import <React/RCTLinkingManager.h>
#import <React/RCTConvert.h>

#if defined(FB_SONARKIT_ENABLED) && __has_include(<FlipperKit/FlipperClient.h>)
#import <FlipperKit/FlipperClient.h>
#import <FlipperKitLayoutPlugin/FlipperKitLayoutPlugin.h>
#import <FlipperKitUserDefaultsPlugin/FKUserDefaultsPlugin.h>
#import <FlipperKitNetworkPlugin/FlipperKitNetworkPlugin.h>
#import <SKIOSNetworkPlugin/SKIOSNetworkAdapter.h>
#import <FlipperKitReactPlugin/FlipperKitReactPlugin.h>

static void InitializeFlipper(UIApplication *application) {
  FlipperClient *client = [FlipperClient sharedClient];
  SKDescriptorMapper *layoutDescriptorMapper = [[SKDescriptorMapper alloc] initWithDefaults];
  [client addPlugin:[[FlipperKitLayoutPlugin alloc] initWithRootNode:application withDescriptorMapper:layoutDescriptorMapper]];
  [client addPlugin:[[FKUserDefaultsPlugin alloc] initWithSuiteName:nil]];
  [client addPlugin:[FlipperKitReactPlugin new]];
  [client addPlugin:[[FlipperKitNetworkPlugin alloc] initWithNetworkAdapter:[SKIOSNetworkAdapter new]]];
  [client start];
}
#endif

@implementation AppDelegate
- (UIInterfaceOrientationMask)application:(UIApplication *)application supportedInterfaceOrientationsForWindow:(UIWindow *)window {
    return [Orientation getOrientation];
  }

- (RCTBridge *)initializeReactNativeApp:(NSDictionary *)launchOptions
{
  RCTBridge *bridge = [[RCTBridge alloc] initWithDelegate:self launchOptions:launchOptions];
  RCTRootView *rootView = [[RCTRootView alloc] initWithBridge:bridge moduleName:@\\"main\\" initialProperties:nil];
  id rootViewBackgroundColor = [[NSBundle mainBundle] objectForInfoDictionaryKey:@\\"RCTRootViewBackgroundColor\\"];
  if (rootViewBackgroundColor != nil) {
    rootView.backgroundColor = [RCTConvert UIColor:rootViewBackgroundColor];
  } else {
    rootView.backgroundColor = [UIColor whiteColor];
  }

  UIViewController *rootViewController = [[EXScreenOrientationViewController alloc] initWithDefaultScreenOrientationMask:UIInterfaceOrientationMaskAllButUpsideDown];
  rootViewController.view = rootView;
  self.window.rootViewController = rootViewController;
  [self.window makeKeyAndVisible];

  return bridge;
}


- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
[RNBranch initSessionWithLaunchOptions:launchOptions isReferrable:YES]; // <-- add this
#if defined(FB_SONARKIT_ENABLED) && __has_include(<FlipperKit/FlipperClient.h>)
  InitializeFlipper(application);
#endif
  
  self.window = [[UIWindow alloc] initWithFrame:[UIScreen mainScreen].bounds];
#if defined(EX_DEV_LAUNCHER_ENABLED)
  EXDevLauncherController *controller = [EXDevLauncherController sharedInstance];
        controller.updatesInterface = [EXUpdatesDevLauncherController sharedInstance];
  [controller startWithWindow:self.window delegate:(id<EXDevLauncherControllerDelegate>)self launchOptions:launchOptions];
#else
  [self initializeReactNativeApp:launchOptions];
#endif

  [super application:application didFinishLaunchingWithOptions:launchOptions];

  return YES;
 }

- (NSArray<id<RCTBridgeModule>> *)extraModulesForBridge:(RCTBridge *)bridge
{
  // If you'd like to export some custom RCTBridgeModules, add them here!
  return @[];
}

- (NSURL *)sourceURLForBridge:(RCTBridge *)bridge {
 #ifdef DEBUG
    #if defined(EX_DEV_LAUNCHER_ENABLED)
  return [[EXDevLauncherController sharedInstance] sourceUrl];
  #else
  return [[RCTBundleURLProvider sharedSettings] jsBundleURLForBundleRoot:@\\"index\\" fallbackResource:nil];
  #endif
 #else
  return [[NSBundle mainBundle] URLForResource:@\\"main\\" withExtension:@\\"jsbundle\\"];
 #endif
}

// Linking API
- (BOOL)application:(UIApplication *)application openURL:(NSURL *)url options:(NSDictionary<UIApplicationOpenURLOptionsKey,id> *)options {
  #if defined(EX_DEV_LAUNCHER_ENABLED)
  if ([EXDevLauncherController.sharedInstance onDeepLink:url options:options]) {
    return true;
  }
  #endif
  return [RCTLinkingManager application:application openURL:url options:options];
}

// Universal Links
- (BOOL)application:(UIApplication *)application continueUserActivity:(nonnull NSUserActivity *)userActivity restorationHandler:(nonnull void (^)(NSArray<id<UIUserActivityRestoring>> * _Nullable))restorationHandler {
return [RNBranch continueUserActivity:userActivity];
}

@end

#if defined(EX_DEV_LAUNCHER_ENABLED)
@implementation AppDelegate (EXDevLauncherControllerDelegate)

- (void)devLauncherController:(EXDevLauncherController *)developmentClientController
    didStartWithSuccess:(BOOL)success
{
  developmentClientController.appBridge = [self initializeReactNativeApp:[EXDevLauncherController.sharedInstance getLaunchOptions]];
}

@end
#endif
"
`;

exports[`addToPodFile modifies the PodFile file for react-native-orientation-plugin 1`] = `
"require File.join(File.dirname(\`node --print \\"require.resolve('expo/package.json')\\"\`), \\"scripts/autolinking\\")
require File.join(File.dirname(\`node --print \\"require.resolve('react-native/package.json')\\"\`), \\"scripts/react_native_pods\\")
require File.join(File.dirname(\`node --print \\"require.resolve('@react-native-community/cli-platform-ios/package.json')\\"\`), \\"native_modules\\")

platform :ios, '12.0'

require 'json'
podfile_properties = JSON.parse(File.read('./Podfile.properties.json')) rescue {}

target 'TestApplication' do
  use_expo_modules!
  config = use_native_modules!

  pod 'expo-dev-launcher', path: '../node_modules/expo-dev-launcher', :configurations => :debug
  pod 'expo-dev-menu', path: '../node_modules/expo-dev-menu', :configurations => :debug
  pod 'react-native-orientation', path: '../node_modules/react-native-orientation'
  use_react_native!(
    :path => config[:reactNativePath],
    :hermes_enabled => podfile_properties['expo.jsEngine'] == 'hermes'
  )

  # Uncomment to opt-in to using Flipper
  #
  # if !ENV['CI']
  #   use_flipper!('Flipper' => '0.75.1', 'Flipper-Folly' => '2.5.3', 'Flipper-RSocket' => '1.3.1')
  # end

  post_install do |installer|
    react_native_post_install(installer)

    # Workaround \`Cycle inside FBReactNativeSpec\` error for react-native 0.64
    # Reference: https://github.com/software-mansion/react-native-screens/issues/842#issuecomment-812543933
    installer.pods_project.targets.each do |target|
      if (target.name&.eql?('FBReactNativeSpec'))
        target.build_phases.each do |build_phase|
          if (build_phase.respond_to?(:name) && build_phase.name.eql?('[CP-User] Generate Specs'))
            target.build_phases.move(build_phase, 0)
          end
        end
      end
    end
  end

end
"
`;
